### PR TITLE
[WIP] Migrate to CRDs

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
- [X]  Vendor CRD base cluster api
- [X] Make code changes to use new cluster-api and build it.
- [X] Local testing to verify if machine object creation using kubectl results into spawning of libvirt instance
- [X] Local testing to verify if deleting machine object, triggers libvirt instance deletion
- [x] Handle `LibvirtMachineProviderStatus`
- [ ] Fix Makefile and Dockerfile 
- [ ] Fix test cases
- [x] Update machine object status

command to build at the moment:
> go build -o bin/manager github.com/openshift/cluster-api-provider-libvirt/cmd/manager

### Local testing is done using these steps:
1. Launch cluster using minikube
2. Create cluster-api-bootstarp.yml
```
[vikas@fed-node cluster-api-provider-libvirt]$ kustomize build vendor/sigs.k8s.io/cluster-api/config/default/ >> /home/vikas/cluster-api-bootstrap.yaml
```
3. kubectl apply -f  /home/vikas/cluster-api-bootstrap.yaml
4. Run libvirt actuator binary:
```
bin/manager --kubeconfig=/root/.kube/config --alsologtostderr
```
5. Create a cluster object using kubectl
```
apiVersion: cluster.k8s.io/v1alpha1
kind: Cluster
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cluster-sample
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - "167.1.1.0/24"
    services:
      cidrBlocks:
      - "70.1.1.0/24"
    serviceDomain: "abc.com"
```
6. Create a machine object using kubectl
```
---
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: machine-sample2
spec:
  providerConfig:
    ValueFrom: null
    value:
      apiVersion: libvirtproviderconfig/v1alpha1
      autostart: false
      domainMemory: 2048
      domainVcpu: 2
      ignKey: "/var/lib/libvirt/images/worker.ign"
      kind: LibvirtMachineProviderConfig
      networkInterfaceAddress: 192.168.124.0/24
      networkInterfaceName: tes
      uri: qemu+tcp://192.168.122.1/system
      volume:
        baseVolumeID: "/var/lib/libvirt/images/coreos_base"
        poolName: default
  versions:
    kubelet: ''
```
7. Verify results using virsh commands.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -15,19 +15,19 @@
...
+# Copy the controller-manager into a thin image
 
 # Final container
 FROM openshift/origin-base
-RUN yum install -y ca-certificates libvirt-libs openssh-clients
-
-COPY --from=builder /go/bin/machine-controller /go/bin/controller-manager /
+WORKDIR /root/
+COPY --from=builder /go/src/github.com/openshift/cluster-api-provider-libvirt/manager .
+ENTRYPOINT ["./manager"]
```
Comment (paulfantom) [inline Dockerfile]: Why we are adding entrypoint?

File: Dockerfile [hunk 2]
```diff
@@ -15,19 +15,19 @@
...
-RUN GOPATH=/go CGO_ENABLED=1 go install ./cmd/machine-controller
-RUN GOPATH=/go CGO_ENABLED=0 GOOS=linux go install -a -ldflags '-extldflags "-static"' github.com/openshift/cluster-api-provider-libvirt/vendor/sigs.k8s.io/cluster-api/cmd/controller-manager
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager github.com/openshift/cluster-api-provider-libvirt/cmd/manager
+
+# Copy the controller-manager into a thin image
 
 # Final container
 FROM openshift/origin-base
-RUN yum install -y ca-certificates libvirt-libs openssh-clients
```
Comment (paulfantom) [inline Dockerfile]: Aren't those libraries needed to connect  to libvirt hypervisor via `qemu+ssh` method?

File: Dockerfile [hunk 3]
```diff
@@ -15,19 +15,19 @@
 # Reproducible builder image
 FROM openshift/origin-release:golang-1.10 as builder
 
-# Workaround a bug in imagebuilder (some versions) where this dir will not be auto-created.
-RUN mkdir -p /go/src/github.com/openshift/cluster-api-provider-libvirt
+# Copy in the go src
 WORKDIR /go/src/github.com/openshift/cluster-api-provider-libvirt
+COPY pkg/    pkg/
```
Comment (paulfantom) [inline Dockerfile]: What does it do?

File: Dockerfile (hunk 1 above)
Comment (vikaschoudhary16) [reply Dockerfile]: There are still parts to be cleaned up. Dockerfile nad Makefile i mentioned explicitly in description. Will cleanup soon.

File: Makefile [hunk 4]
```diff
@@ -1,69 +1,51 @@
...
+manifests:
+	go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go all
+
+# Run go fmt against code
+fmt:
+	go fmt ./pkg/... ./cmd/...
+
+# Run go vet against code
+vet:
+	go vet ./pkg/... ./cmd/...
```
Comment (paulfantom) [inline Makefile]: Why this changed from using `hack/go-fmt.sh` script?

File: Makefile [hunk 5]
```diff
@@ -1,69 +1,51 @@
...
+	kubectl apply -f config/crds
+	kustomize build config/default | kubectl apply -f -
+
+# Generate manifests e.g. CRD, RBAC etc.
+manifests:
+	go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go all
+
+# Run go fmt against code
+fmt:
+	go fmt ./pkg/... ./cmd/...
```
Comment (paulfantom) [inline Makefile]: Why this changed from using `hack/go-fmt.sh` script?

File: Makefile [hunk 6]
```diff
@@ -1,69 +1,51 @@
...
-# limitations under the License.
-
-VERSION     ?= $(shell git describe --always --abbrev=7)
-MUTABLE_TAG ?= latest
-IMAGE        = origin-libvirt-machine-controllers
-
-.PHONY: all
-all: build images check
-
-NO_DOCKER ?= 0
```
Comment (paulfantom) [inline Makefile]: Why this is removed?

File: Makefile [hunk 7]
```diff
@@ -1,69 +1,51 @@
...
-push:
-	docker push "$(IMAGE):$(VERSION)"
-	docker push "$(IMAGE):$(MUTABLE_TAG)"
-
-.PHONY: check
-check: fmt vet lint test ## Check your code
-
-.PHONY: test
-test: # Run unit test
-	$(DOCKER_CMD) go test -race -cover ./cmd/... ./cloud/...
```
Comment (paulfantom) [inline Makefile]: Running go should be containerized to ensure local test env is identical with CI one.

Comment (frobware) [reply Makefile]: > Running go should be containerized to ensure local test env is identical with CI one.

I think our Makefile's are upside down. They should favour the developer who types a lot of make commands. In dev, I juat want a local binary.

Comment (paulfantom) [reply Makefile]: Produced binary is preserved in `bin/` directory and our makefiles allow to use non-containerized builds.

**Comment by paulfantom** [comment]
Shouldn't this PR use `crd` as target branch instead of `master`?

**Comment by enxebre** [comment]
@vikaschoudhary16 please shout when this is ready for review :)
can you please before rebase into reasonable commits? this is the ones we used on aws fo consistency
- revendor for CRDs
- add kubebuilder generated skeleton
- drop old machine controller and build new one
- move provider config for new layout
- update actuator for using new provider config
- update docker and Makefile
- fix lints
Previously we suited the layout on master moving the cloud folder into packages
cc @ingvagabund

**Comment by ingvagabund** [comment]
Closing in favor of https://github.com/openshift/cluster-api-provider-libvirt/pull/72

**Comment by vikaschoudhary16** [comment]
oh, this was very aggressive overtake. 
1. It was holiday in India on Friday
2. I was not well throughout week but was still planning to work on this during weekend that i told Alberto too.

**Comment by ingvagabund** [comment]
Sorry about that Vikas :( Alberto restructured his PR during the week with additional changes that we discussed, the e2e framework has changed, some downstream patches were applied. I did not want you to spend time duplicating the same job we did with Alberto.

**Pull request closed without merging**